agg-ast/schema_derivation/src/lib.rs

```diff
@@ -576,3 +576,113 @@ macro_rules! array_element_schema_or_error {
         }
     }};
 }
+
+/// get_namespaces_for_pipeline is a helper function that takes a pipeline and the
+/// database and collection it is called on, and iterates through the stages, parsing
+/// out all collection namespaces referenced in the pipeline. This allows for more
+/// concise catalogs to be sent to derive_schema_for_pipeline, rather than the catalog
+/// for the entire database.
+pub fn get_namespaces_for_pipeline(
+    pipeline: Vec<agg_ast::definitions::Stage>,
+    current_db: String,
+    current_collection: Option<String>,
+) -> BTreeSet<agg_ast::definitions::Namespace> {
+    let mut namespaces = BTreeSet::new();
+
+    // because a pipeline references collections from within the same database,
+    // we can use that database to create a namespace and add it to the set. This
+    // macro makes it easier to read for places where we unpack a collection name.
+    macro_rules! add_namespace {
+        ($coll:expr) => {
+            namespaces.insert(agg_ast::definitions::Namespace::new(
+                current_db.clone(),
+                $coll,
+            ));
+        };
+    }
+
+    // if the current agg pipeline is on a collection (i.e. not aggregate 1), add that
+    // namespace to the set
+    if let Some(current_collection) = current_collection {
```

Comment:
This will make it easier to use later, if needed.
```suggestion
    if let Some(current_collection) = current_collection.as_ref() {
```